reps/2024-06-13-export-api.md

```diff
@@ -0,0 +1,141 @@
+and stopping of a job, the constuction and destruction of an Actor, etc. Event is structured data,
+and each resource corresponds to a specified event type. All types of events follow a generally consistent format,
+but the event fields of different types can be customized. E.g.:
+
+```json
+{"event_type": "JOB", "job": {"submission_id": "raysubmit_WBMEB9nTaMKmrKrN", "job_info": {"status": "RUNNING", "entrypoint": "python3 my_script.py", "message": null, "error_type": null, "start_time": 1692004435106, "end_time": null, "metadata": null, "entrypoint_num_cpus": null, "entrypoint_num_gpus": null, "entrypoint_resources": null, "driver_agent_http_address": null, "driver_node_id": null}}}
+
+{"event_type": "ACTOR", "actor": {"actor_id": "efc87749b2e337d7872dcf3802000000", "actor_info": {"actorId": "efc87749b2e337d7872dcf3802000000", "jobId": "02000000", "address": {"rayletId": "ffffffffffffffffffffffffffffffffffffffffffffffffffffffff", "workerId": "ffffffffffffffffffffffffffffffffffffffffffffffffffffffff", "ipAddress": "", "port": 0}, "name": "_ray_internal_job_actor_raysubmit_WBMEB9nTaMKmrKrN", "className": "JobSupervisor", "state": "PENDING_CREATION", "numRestarts": "0", "timestamp": 0.0, "pid": 0, "startTime": 0, "endTime": 0, "actorClass": "JobSupervisor", "exitDetail": "-", "requiredResources": {}}}}
+
+```
+
+By aggregating all the Events in time dimension, we can obtain the historical state of the ray cluster,
+and the consumers of the events can aggregate the final state from a series of events. The export api
+does not guarantee the ultimate consistency of event state, which is feasible and tolerated in production practice.
+The loss of some intermediate events will not cause the final state to be incorrect. For example,
+if the event of an actor's creation is lost, but the consumer can read the event of actor exiting,
+then the consumer knows that the actor finally ran successfully.
+
+
+The various components of Ray, including GCS, Dashboard, raylet, and worker, can all serve as event generators.
+However, these sources don't need to concern themselves with the details of how events are stored, transmitted, or published.
+
+### Events to Export and How
+#### Events to Export
+- Core
+Tasks/Actors/Objects/Jobs/Nodes/Placement groups, including its meta data and current status when event emits.
+
+|  Event            | Event Source          | When to export                                                                    | Format example(maybe a file link)|
+|  ----             | ----                  | ----                                                                              | ----                             |
+| Tasks             | CoreWorker and raylet | When coreworker adds task event to<br>ray::core::worker::TaskEventBuffer          |          |
+| Actors            | GCS                   | GcsPublisher::PublishActor<br>publish actor status through<br>GCS_ACTOR_CHANNEL   |          |
```

Comment:
Schema for actors: https://docs.google.com/document/d/1_7Z81KtmjA3IAEeZqEuxvw6ur5acyU-CrofQ6rFEmTg/edit